[ctp] New port CTP library 6.3.15

xgao1023 opened the pull request:
The Comprehensive Transaction Platform (CTP) is a future brokerage management system developed specially for futures companies. CTP API client library allows users to connect to CTP gateway for algorithmic trading.

The API client library is pre-built and only support Windows (x86, x64) and Linux (x64 only)

Commit:
[ctp] New port CTP library 6.3.15

xgao1023 commented:
How should I disable the CI tests for the platforms not supported? 
The library is pre-built (not open-source) and only supports limited platforms.

JackBoosY commented:
Thanks for your PR, but vcpkg only supports installation packages to install binaries(eg: CUDA), other ports should be built from source.

xgao1023 commented:
> Thanks for your PR, but vcpkg only supports installation packages to install binaries(eg: CUDA), other ports should be built from source.

Thanks! But I'm still a bit confused.
Does it mean 
1. there is no proper way to introduce closed-source libraries into vcpkg (CUDA is a special case)?
2. Or; there is a proper way that I'm not aware of?

JackBoosY commented:
@xgao1023 Yes, only SDKs containing installation packages will be allowed to be added to the vcpkg port list (eg CUDA, opengl, zedsdk and vulkan). The reason why vcpkg requires the use of source code to build binaries is that it is more secure and more open source.

JackBoosY commented:
/azp run

JackBoosY commented:
@xgao1023 Well, after discussion, we think we can now accept this PR.
So please resolve these regressions.

Thanks.

xgao1023 commented:
> @xgao1023 Well, after discussion, we think we can now accept this PR.
> So please resolve these regressions. Logs link [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=36815&view=artifacts&type=publishedArtifacts).
> 
> Thanks.

Hi,

Glad to hear that it's possible to add such pre-built binary library to vcpkg.

However, this library is only available for x86-windows, x64-windows and x64-linux. How should I disable (or skip) regression of other platforms?

Thanks

JackBoosY commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,68 @@
+message(STATUS "The downloading server may only be avilable during 8:00AM-5:00PM (GMT+8)")
+
```
Add the following code here:
```
vcpkg_fail_port_install(ON_TARGET "OSX" "UWP" ON_ARCH "arm")
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```

xgao1023 replied:
Thanks!

But I'm thinking probably I should use 
`vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)`

Because the library only provide dynamic shared library version (DLL with import lib on Windows and .so on Linux)

I haven't checked if it works with static CRT library - will confirm.

JackBoosY replied:
@xgao1023 Yes, you're right.

JackBoosY reviewed (changes requested):
Please add the following code to _VCPKG_PATH/scripts/ci.baseline.txt_:
```
ctp:x64-windows-static=fail
ctp:arm64-windows=fail
ctp:arm-uwp=fail
ctp:x64-uwp=fail
ctp:x64-osx=fail
```

xgao1023 commented:
/azp run

xgao1023 commented:
Hi,

I've address the changes related to regression. But I think the CI jobs fail for arm and arm64 at the step of initializing environment.

Can check and help?

Thanks

strega-nil commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+message(STATUS "Note: the downloading server may only be avilable during 8:00AM-5:00PM (GMT+8)")
```
```suggestion
message(STATUS "Note: the downloading server may only be available from 8:00AM-5:00PM (UTC+8)")
```

xgao1023 replied:
Well, IMO there is no real good reason, and the available hours are never documented.
In other hours the server might or might not be available - depending on something we don't know.

xgao1023 replied:
Do you know how to disalbe CI for this library?

strega-nil replied:
yeah, just `ctp:<triplet>=skip`, and just do that for every triplet.

xgao1023 replied:
I've already changed accordingly.

strega-nil commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+    NO_REMOVE_ONE_LEVEL
+)
+
+if(${TARGET_TRIPLET} STREQUAL "x86-windows")
```
this should check for `VCPKG_TARGET_ARCHITECTURE` and `VCPKG_TARGET_IS_<target>`, in order to support custom triplets:

```cmake
if(VCPKG_TARGET_IS_WINDOWS AND VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
  ...
elseif(VCPKG_TARGET_IS_WINDOWS AND VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
  ...
elseif(VCPKG_TARGET_IS_LINUX AND VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
  ...
else()
  ...
endif()
```

xgao1023 replied:
Thanks will do it

strega-nil reviewed (changes requested):
Some minor changes, but this looks good. I'm kind of concerned about the downloading server only being available during working hours Chinese time; is there a good reason for that? We also probably want to skip CI always for this library, if it's going to fail at random times.

Commit:
[ctp] Minor changes
 - Update portfile to support custom triplets
 - disable CI due to downloading server availability

JackBoosY commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+
+file(INSTALL ${SOURCE_PATH}/ DESTINATION ${CURRENT_PACKAGES_DIR}/include FILES_MATCHING PATTERN "*.h")
+
+if(${VCPKG_TARGET_IS_WINDOWS})
```
```suggestion
if(VCPKG_TARGET_IS_WINDOWS)
```

JackBoosY commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+    file(INSTALL ${SOURCE_PATH}/ DESTINATION ${CURRENT_PACKAGES_DIR}/bin FILES_MATCHING PATTERN "*.dll")
+    file(INSTALL ${SOURCE_PATH}/ DESTINATION ${CURRENT_PACKAGES_DIR}/Debug/lib FILES_MATCHING PATTERN "*.lib")
+    file(INSTALL ${SOURCE_PATH}/ DESTINATION ${CURRENT_PACKAGES_DIR}/Debug/bin FILES_MATCHING PATTERN "*.dll")
+elseif(${VCPKG_TARGET_IS_LINUX})
```
```suggestion
elseif(VCPKG_TARGET_IS_LINUX)
```

JackBoosY committed:
Update ports/ctp/portfile.cmake

JackBoosY committed:
Update ports/ctp/portfile.cmake

strega-nil commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+    )
+    file(INSTALL ${SOURCE_PATH_DATACOLLECT}/ DESTINATION ${CURRENT_PACKAGES_DIR}/include FILES_MATCHING PATTERN "*.h")
+
+    if(${VCPKG_TARGET_IS_WINDOWS})
```
```suggestion
    if(VCPKG_TARGET_IS_WINDOWS)
```

strega-nil commented on ports/ctp/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+        file(INSTALL ${SOURCE_PATH_DATACOLLECT}/ DESTINATION ${CURRENT_PACKAGES_DIR}/bin FILES_MATCHING PATTERN "*.dll")
+        file(INSTALL ${SOURCE_PATH_DATACOLLECT}/ DESTINATION ${CURRENT_PACKAGES_DIR}/Debug/lib FILES_MATCHING PATTERN "*.lib")
+        file(INSTALL ${SOURCE_PATH_DATACOLLECT}/ DESTINATION ${CURRENT_PACKAGES_DIR}/Debug/bin FILES_MATCHING PATTERN "*.dll")
+    elseif(${VCPKG_TARGET_IS_LINUX})
```
```suggestion
    elseif(VCPKG_TARGET_IS_LINUX)
```

strega-nil committed:
Remove ${...} from if

strega-nil commented:
Thanks so much @xgao1023 ! This is a great port :)
